src/Algebra/Solver/IdempotentCommutativeMonoid/Normal.agda

```diff
@@ -0,0 +1,157 @@
+--
+-- Adapted from Algebra.Solver.CommutativeMonoid
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+open import Algebra.Bundles using (IdempotentCommutativeMonoid)
+
+module Algebra.Solver.IdempotentCommutativeMonoid.Normal
+  {c ℓ} (M : IdempotentCommutativeMonoid c ℓ) where
+
+import Algebra.Properties.CommutativeSemigroup as CSProperties
+open import Data.Bool as Bool using (Bool; true; false; if_then_else_; _∨_)
+open import Data.Fin.Base using (Fin; zero; suc)
+open import Data.Nat.Base using (ℕ)
+open import Data.Vec.Base using (Vec; []; _∷_; lookup; replicate; zipWith)
+import Data.Vec.Relation.Binary.Pointwise.Inductive as Pointwise
+open import Relation.Binary.Definitions using (DecidableEquality)
+import Relation.Binary.Reasoning.Setoid as ≈-Reasoning
+import Relation.Nullary.Decidable       as Dec
+
+open IdempotentCommutativeMonoid M
+open CSProperties commutativeSemigroup using (interchange; x∙yz≈xy∙z; x∙yz≈y∙xz)
+open ≈-Reasoning setoid
+
+private
+  variable
+    n : ℕ
+
+
+------------------------------------------------------------------------
+-- Monoid expressions
+
+open import Algebra.Solver.Monoid.Expression rawMonoid public
+  hiding (NormalAPI)
+
+------------------------------------------------------------------------
+-- Normal forms
+
+-- A normal form is a vector of bits (a set).
+
+Normal : ℕ → Set
+Normal n = Vec Bool n
+
+-- The semantics of a normal form.
+
+⟦_⟧⇓ : Normal n → Env n → Carrier
+⟦ []    ⟧⇓ _ = ε
+⟦ b ∷ v ⟧⇓ (a ∷ ρ) = if b then a ∙ (⟦ v ⟧⇓ ρ) else (⟦ v ⟧⇓ ρ)
+
+-- We can decide if two normal forms are /syntactically/ equal.
+
+infix 5 _≟_
+
+_≟_ : DecidableEquality (Normal n)
+nf₁ ≟ nf₂ = Dec.map Pointwise-≡↔≡ (decidable Bool._≟_ nf₁ nf₂)
+  where open Pointwise using (Pointwise-≡↔≡; decidable)
+
+------------------------------------------------------------------------
+-- Constructions on normal forms
+
+-- The empty set.
+
+empty : Normal n
+empty = replicate _ false
+
+-- A singleton set.
+
+sg : (i : Fin n) → Normal n
+sg zero    = true ∷ empty
+sg (suc i) = false ∷ sg i
+
+-- The composition of normal forms.
+infixr 10 _•_
+
+_•_ : (v w : Normal n) → Normal n
+_•_ = zipWith _∨_
+
+------------------------------------------------------------------------
+-- Correctness of the constructions on normal forms
+
+-- The empty set stands for the unit ε.
+
+empty-correct : (ρ : Env n) → ⟦ empty ⟧⇓ ρ ≈ ε
+empty-correct []      = refl
+empty-correct (a ∷ ρ) = empty-correct ρ
+
+-- The singleton set stands for a single variable.
+
+sg-correct : (x : Fin n) (ρ : Env n) → ⟦ sg x ⟧⇓ ρ ≈ lookup ρ x
+sg-correct zero (x ∷ ρ) = begin
+    x ∙ ⟦ empty ⟧⇓ ρ   ≈⟨ ∙-congˡ (empty-correct ρ) ⟩
+    x ∙ ε              ≈⟨ identityʳ _ ⟩
+    x                  ∎
+sg-correct (suc x) (m ∷ ρ) = sg-correct x ρ
+
+-- Normal form composition corresponds to the composition of the monoid.
+
+distr : ∀ a b c → a ∙ (b ∙ c) ≈ (a ∙ b) ∙ (a ∙ c)
```

Comment:
This feels like it should be a lemma in `Properties.IdempotentCommutativeMonoid`?

Reply:
It is now, but the PR that introduced that was submitted *after* this PR, but *merged* before... ;-)